### PR TITLE
[WIP] Add exception filter rewriter for AOT scenarios

### PR DESCRIPTION
**Pull request by kg**
This PR will add a tool that rewrites exception filters into non-filtered code so we can get code with filters working in fullaot, then integrate it into the wasm packager flow.

### CODE REVIEW

File: mcs/tools/exception-filter-rewriter/Program.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                foreach (var arg in _args)
+                    ParseArgument(arg, options);
+
+                var argv = _args.Where(arg => !arg.StartsWith("-")).ToArray();
+
+                var step = options.Overwrite ? 1 : 2;
+                if (argv.Length < step) {
+                    Usage();
+                    return 1;
+                }
```
Comment (lambdageek) [inline mcs/tools/exception-filter-rewriter/Program.cs]: Looks like most of this file isn't formatted to the [Mono Coding Guidelines](https://www.mono-project.com/community/contributing/coding-guidelines/) - should be tabs (width=8) not spaces; and braces for methods go on a new line; space before `(` of a method call; etc.

**Comment by MichalStrehovsky** [comment]
Out of curiosity - why isn't this done in the LLVM importer?

IL rewriting is what we did in .NET Native and it's the number one reason why the compilation throughput of .NET Native is so bad. We applied the lessons learned in CoreRT and the CoreRT AOT compiler is like 10x faster than .NET Native (mostly thanks to getting rid of IL rewriting).

**Comment by kg** [comment]
> Out of curiosity - why isn't this done in the LLVM importer?
> 
> IL rewriting is what we did in .NET Native and it's the number one reason why the compilation throughput of .NET Native is so bad. We applied the lessons learned in CoreRT and the CoreRT AOT compiler is like 10x faster than .NET Native (mostly thanks to getting rid of IL rewriting).

Doing it in the LLVM importer sounds like a great long-term solution, but my understanding is that expressing exception filters in LLVM is very complex. I guess we could do this full rewrite at an IL level during ingest before the AOT generation happens, but that seems like it introduces high debugging risk. With this approach I can verify the rewritten binary on windows netframework, netcore, and mono, along with running it through peverify and dnspy. (this has been quite useful)

**Comment by MichalStrehovsky** [comment]
> Doing it in the LLVM importer sounds like a great long-term solution, but my understanding is that expressing exception filters in LLVM is very complex

I meant doing this transformation at the time LLVM trees are built (not within LLVM). LLVM doesn't have a model for .NET two pass exception handling so all that can be done is an approximation like in this pull request and hope customer code doesn't rely on filters running before the nested finally blocks.

> With this approach I can verify the rewritten binary on windows 

Yeah, that was one of the reasons why we did this initially for .NET Native too. Eventually we ended up structuring the compiler in a way that doesn't hinder developer productivity, but also doesn't require customers to suffer roundtripping assemblies to disk in retail scenarios.

(E.g. the compiler now allows dumping modified IL in text form to disk and mapping debug information to it to enable source stepping in it, should one need that.)

**Comment by lewurm** [comment]
FWIW this is the proposal to implement exception filters via funclets https://github.com/mono/mono/issues/8941

I would love to see some tests, and maybe some examples how the rewritten result looks like.

**Comment by kg** [comment]
I accidentally pushed this to mono/mono (oops) so I need to make a new PR. Closing

**Pull request closed without merging**